### DISCUSSION TITLE
Update Function Choice Behavior for a chat completion function

### DISCUSSION THREAD

**Discussion by HuskyDanny** [microsoft/semantic-kernel#9660] [General]
Hi SK Team,

I added the function with the req setting and prompt template as default:
![image](https://github.com/user-attachments/assets/7a749906-47d1-4a02-9cdf-0261c4f1ac04)
Now I want to update the req_setting when calling to this:
![image](https://github.com/user-attachments/assets/bc4d629a-b417-4d09-b7da-f67c0cc289cb)
What I have tried it to pass in the arguments for kernel.invoke
![image](https://github.com/user-attachments/assets/8a6ee072-2bc6-400c-b756-6ffa92780276)
But the result is weird, the included function is executed but the qa_with_context chat completion function itself is only returning empty string, Is there something wrong? Things works fine with the default, but I want the plug in to always get invoked so I update to the new one.
Here is the debug snapshot of the returned response
![image](https://github.com/user-attachments/assets/98a158f8-2e82-4d23-8646-e763ff293ce7)
here is the request snapshot:
![image](https://github.com/user-attachments/assets/fc358061-3c30-4b13-b25f-ff2e8184cb28)

**Comment by moonbox3** [comment]
Hi @HuskyDanny, I am looking into this further and I am running the following Python sample:

https://github.com/microsoft/semantic-kernel/blob/main/python/samples/concepts/auto_function_calling/function_calling_with_required_type.py

The FunctionChoiceBehavior settings have the type set as `Required`. As we covered, this means the model will call one of the advertised tools. Once the `function_choice_behavior.maximum_auto_invoke_attempts` are hit, we get into the following code:

https://github.com/microsoft/semantic-kernel/blob/7579b243f4f428d984e9d99300dde7256eb85a06/python/semantic_kernel/connectors/ai/chat_completion_client_base.py#L171

This code clears the advertised tools and tool choice so that the model, using its Chat History, can respond with a final answer. In the sample code I linked to above, the `maximum_auto_invoke_attempts` are configured as 1. The model can request to run one or both of the advertised tools, we then run those functions in SK, and provide the results back to the model to form its final answer.
